### DISCUSSION TITLE
Does calling choices/macros from an inline script return anything?

### DISCUSSION THREAD

**Discussion by JulianBroudy** [chhoumann/quickadd#392] [Q&A]
### Motivation:
-  **I would like to have** the option to choose which parameters I want when asked to add a task. 
- **I don't want to have** 30+ capture options from which I have to choose which one I want this time.

### Goal:
- [ ] Create 5 captures, each for 1 needed field with `{{DATE:<DATEFORMAT>}}`. In another choice add inline script to call them based on the 1st checkbox prompt.

### Current Solution:
- [x] Create inline script that gives a checkbox and per each checked box get the appropriate input.

### The only option I found to do this is:

1. Create a choice with inline script.
2. Use the following script:
```js
const description = await this.quickAddApi.inputPrompt("Enter task description:");
let output = description;

const options = ["Priority", "Due Date", "Scheduled Date", "Start Date", "Recurrence"];
const selectedOptions = await this.quickAddApi.checkboxPrompt(options, []);

const nldates = this.app.plugins.getPlugin("nldates-obsidian");

if (selectedOptions.includes("Due Date")) {
    const dueDate = await this.quickAddApi.inputPrompt("Enter due date:");
    output += ` 📅 ` + nldates.parseDate(dueDate).formattedString;
}
if (selectedOptions.includes("Scheduled Date")) {
    const scheduledDate = await this.quickAddApi.inputPrompt("Enter scheduled date:");
    output += ` ⏳ ` + nldates.parseDate(scheduledDate ).formattedString;
}
if (selectedOptions.includes("Start Date")) {
    const startDate = await this.quickAddApi.inputPrompt("Enter start date:");
    output += ` 🛫 ` + nldates.parseDate(startDate ).formattedString;
}
if (selectedOptions.includes("Recurrence")) {
    const recurrence = await this.quickAddApi.inputPrompt("Enter recurrence:");
    output += ` 🔁 ` + nldates.parseDate(recurrence ).formattedString;
}
return output;
```

**Comment by chhoumann** [comment]
Hi @JulianBroudy 

As for your question, only macros will return anything when placed in a capture format. So if I have `{{MACRO:script.js}}`, and the script returns `test`, then `{{MACRO:script.js}}` would be replaced by `test` in the final output.
Likewise, you could export multiple functions from your script and call them with, e.g., `{{MACRO:script.js::exportedFn}}`.
Note that I've called the macro `script.js`.

Personally, I think your current solution is pretty good. Is there a particular reason why you want to separate the script into captures, and then call them from the script?

**Reply by JulianBroudy** [reply]
Hi @chhoumann, thank you for the prompt reply. 

Mainly because this way I only have the option to get the checkbox. 

If I have multiple captures each takes a specific variable like `due date` or `start date` then I can create a Macro with 2 or 3 mostly used options. As in I mostly use `due date` alone, or `scheduled date` with `occurrence`, or etc. And have this script when I need some other combination.

But I was unable to make it dynamic through the built-in functions (correct me if I'm wrong or missed anything). So, I opted out for the above solution. (💡 Maybe it would be possible to create built-in conditional choices flow in a Macro.) 

Looking [here](https://quickadd.obsidian.guide/docs/QuickAddAPI#executechoicechoicename-string-variables-key-string-any-promise), I am unsure whether calling a choice from inline script would return anything. Does it?

Thank you again for your input.

**Reply by chhoumann** [reply]
That makes sense.

As for your idea: https://github.com/chhoumann/quickadd/issues/18
I'm currently thinking about improvements to macros that I want to implement soon. This is one of them.

Calling `executeChoice` does not return anything. It returns `Promise<void>`.

I guess you could create a macro with just a script acting as a controller of sorts, export multiple functions from that script file (for code-reuse), and then using those in the captures. After all, calling a macro from a capture (like before, e.g. `{{MACRO:script.js::exportedFn}}`) can influence the variables accessible through `{{VALUE:...}}`.

For example, this may be of use:
```js
module.exports = {
    exportedFn: async function (params) {
        params.variables["test"] = "test";
        return `test`;
    },
};
```

The script sets the `{{VALUE:test}}` variable to `test`.
In the capture, my format is:
```markdown
{{MACRO:script.js::exportedFn}}
{{VALUE:test}}
```
which evaluates to
```markdown
test
test
```
